lambdas/data-migration2/src/granulesAndFiles.ts

```diff
@@ -76,8 +76,13 @@ export const migrateGranuleRecord = async (
     }
   }
 
-  // Throw error if it was already migrated.
-  if (existingRecord) {
+  const isExistingRecordNewer = existingRecord
+    && existingRecord.updated_at >= new Date(record.updatedAt);
+
+  // Check the record's status here to prevent a granule from being written
+  // out-of-order. If we have already migrated this granule and the status
+  // of the new granule is 'running', skip it.
+  if (isExistingRecordNewer || (existingRecord && record.status === 'running')) {
```

Comment:
I'm not 100% sure here but I think this is what we want to preserve the existing upsert logic where we're guarding against out-of-order writes.

This means that we will never update a record in the `running` state. That seems safest given the concerns but I want to flag it. 

An alternative might be to add `updated_at` to the `GranulesExecutions` JOIN table. We could check that date and if it's less than or equal to the dynamo record's updated_at, we could update:

`if (granulesExecutionsRecord.updated_at <= dynamoRecord.updated_at) { update the granule and granulesExecution }`

Reply:
Can you help me understand what we're trying to defend against here? For the parallel writes, we need the protection against out-of-order writes because the events can be received out of order. 

But the migrations should just take whatever is in Dynamo and copy it to PG, regardless of what status it is in or otherwise, unless I'm missing something.

There will be cases where at the time we migrate a granule, it is "running" in Dynamo, so copies to PG as "running". But then the execution finishes and the granule becomes "completed" in Dynamo and it **should** get parallel written to Dynamo and PG

Reply:
Yeah, so the issue here is the `upsert` logic in `GranulePgModel`. 

If I remove this line in the data-migration lambda that checks the status, we'll always try to upsert a granule if the record coming in is newer than the one already migrated. That's a problem because the `GranulePgModel.upsert` logic won't actually allow the record to be updated if it's in the `running` status. In that case we'll already have a `granulesExecutionsPgRecord` so we'll trigger the `whereNotExists` clause and prevent the upsert. 

Does that make sense? Happy to talk it through.

Reply:
Summing up our side conversation: we'll skip this `status` check in data-migration2 and handle the error that comes when an upsert fails due to a running granule.

Reply:
This commit:

https://github.com/nasa/cumulus/pull/2162/commits/6bead8bf8780e996500c39e0df452b90788f69f7

throws a new error type if the upsert didn't return any rows and we were therefor not able to update a granule during migration.